### PR TITLE
feat(tf/pt): add/refact lammps & C++ support for spin model

### PR DESCRIPTION
**Pull request by iProzd**
The conversations in #4216 are not readable. I opened this clean one for reviewers.

## Release Notes

- **New Features**
  - Introduced support for spin configurations within the DeepMD framework.
  - Added `PairDeepSpin` class for integrating spin models into LAMMPS simulations.
  - Enhanced `DeepSpin` and `DeepSpinModelDevi` classes with new methods for energy and force calculations.
  - Updated `DeepPotTF` class to compute atomic energies and virials.
  - Added a new function for writing LMP data that includes spin information.

- **Bug Fixes**
  - Improved error handling for unsupported configurations in spin calculations.

- **Tests**
  - Added comprehensive unit tests for `DeepSpin` and `DeepSpinModelDevi` classes to validate functionality and ensure accuracy in computations.
  - Introduced tests for verifying spin interactions without periodic boundary conditions.
  - Developed tests for MPI-enabled versions of spin interactions.

- **Documentation**
  - Updated scripts and tests to reflect new functionalities and ensure clarity in usage.

### CODE REVIEW

**Comment by iProzd** [comment]
> > The conversations in #4216 are not readable. I opened this clean one for reviewers.
> 
> I found that my comment in #4216 hasn't been resolved

Is this [comment](https://github.com/deepmodeling/deepmd-kit/pull/4216/#discussion_r1801776864)? I found that another PR during this one has already bumped the c_api version. Fixed in https://github.com/deepmodeling/deepmd-kit/pull/4321/commits/e572e376c93f1095d84552e4516048c1906fe3b5.

File: source/api_cc/src/DeepPotTF.cc [hunk 1]
```diff
@@ -1026,4 +1034,12 @@ void DeepPotTF::computew_mixed_type(std::vector<double>& ener,
...
                      coord, atype, box, fparam, aparam, atomic);
 }
+
+void DeepPotTF::cum_sum(std::map<int, int>& sum, std::map<int, int>& vec) {
+  sum[0] = 0;
+  for (int ii = 1; ii < vec.size(); ++ii) {
+    sum[ii] = sum[ii - 1] + vec[ii - 1];
+  }
+}
+
```
Comment (njzjz) [inline source/api_cc/src/DeepPotTF.cc]: I don't get the reason to add a class method here.

File: source/lmp/pair_deepmd.cpp [hunk 2]
```diff
@@ -479,18 +120,10 @@ void PairDeepMD::compute(int eflag, int vflag) {
...
-    // get spin
-    for (int ii = 0; ii < nall; ++ii) {
-      for (int dd = 0; dd < 3; ++dd) {
-        dspin[ii * 3 + dd] = sp[ii][dd];
-      }
-    }
+    throw std::runtime_error(
+        "Pair style 'deepmd' does not support spin atoms, please use pair "
+        "style 'deepspin' instead.");
   }
```
Comment (njzjz) [inline source/lmp/pair_deepmd.cpp:127]: Use `error->all`.

File: source/api_c/include/c_api.h [hunk 3]
```diff
@@ -94,6 +94,20 @@ extern void DP_DeleteNlist(DP_Nlist* nl);
  */
 const char* DP_NlistCheckOK(DP_Nlist* dp);
 
+/**
+ * @brief The deep potential base model.
```
Comment (njzjz) [inline source/api_c/include/c_api.h:98]: I suggest adding doc `@since API version 24` to all new functions you added in this PR.

https://github.com/deepmodeling/deepmd-kit/blob/edb1e9fecd448512c87806da97b3732c44a0cc2b/source/api_c/include/c_api.h#L13

File: source/api_c/include/c_api.h [hunk 4]
```diff
@@ -317,6 +365,51 @@ extern void DP_DeepPotCompute2(DP_DeepPot* dp,
...
+ * @param[in] spin The spins of atoms, [0, 0, 0] if no spin. The array should be
+ *of size nframes x natoms x 3.
+ * @param[in] atype The atom types. The array should contain natoms ints.
+ * @param[in] box The cell of the region. The array should be of size 9. Pass
+ *NULL if pbc is not used.
+ * @param[in] fparam The frame parameters. The array can be of size nframes x
+ *dim_fparam.
+ * @param[in] aparam The atom parameters. The array can be of size nframes x
+ *dim_aparam.
+ * @param[out] energy Output energy.
```
Comment (njzjz) [inline source/api_c/include/c_api.h:386]: The size seems to be nframes x natoms x dim_aparam.

File: source/api_c/include/c_api.h [hunk 5]
```diff
@@ -317,6 +365,51 @@ extern void DP_DeepPotCompute2(DP_DeepPot* dp,
...
+ * @param[out] force Output force. The array should be of size natoms x 3.
+ * @param[out] force_mag The magnetic force on each atom.
```
Comment (njzjz) [inline source/api_c/include/c_api.h:388]: Since the array needs to allocate the memory in advance, it's important to document the size.

File: source/api_c/include/c_api.h (hunk 4 above)
Comment (njzjz) [reply source/api_c/include/c_api.h:386]: It seems an old issue, though

**Review by njzjz — CHANGES_REQUESTED** [review]

File: source/api_cc/include/DeepPot.h [hunk 6]
```diff
@@ -687,10 +611,10 @@ class DeepPotModelDevi {
...
   void compute(std::vector<ENERGYTYPE>& all_ener,
-               std::vector<std::vector<VALUETYPE> >& all_force,
-               std::vector<std::vector<VALUETYPE> >& all_virial,
-               std::vector<std::vector<VALUETYPE> >& all_atom_energy,
-               std::vector<std::vector<VALUETYPE> >& all_atom_virial,
+               std::vector<std::vector<VALUETYPE>>& all_force,
+               std::vector<std::vector<VALUETYPE>>& all_virial,
+               std::vector<std::vector<VALUETYPE>>& all_atom_energy,
+               std::vector<std::vector<VALUETYPE>>& all_atom_virial,
                const std::vector<VALUETYPE>& coord,
```
[the remark this answers is not in the document]
Comment (iProzd) [reply source/api_cc/include/DeepPot.h:614]: I found that these are made by pre-commit.

[the remark this answers is not in the document]
Comment (iProzd) [reply source/api_cc/include/DeepPot.h:614]: see https://github.com/deepmodeling/deepmd-kit/pull/4321/commits/388bb22e6fb4bb6509c2a8b98ad39b58e9117277

**Review by njzjz — COMMENTED** [review]
I took a look at the codecov, and found many new lines are not covered: https://app.codecov.io/gh/deepmodeling/deepmd-kit/pull/4321

**Review by wanghan-iapcm — COMMENTED** [review]
please fix the ut coverage issue pointed out by @njzjz

**Review by njzjz — APPROVED** [review]

**Review by CaRoLZhangxy — APPROVED** [review]
